p2p/peer.go

```diff
@@ -235,25 +239,91 @@ type PeerUpdate struct {
 // - Disconnected: peer disconnects, unmarking as connected and broadcasts a
 //   PeerStatusDown peer update.
 //
+// If we need to evict a peer, typically because we have connected to additional
+// higher-scored peers and need to shed lower-scored ones, the flow is as follows:
+// - EvictNext: returns a peer ID to evict, marking peer as evicting.
+// - Disconnected: peer was disconnected, unmarking as connected and evicting,
+//   and broadcasts a PeerStatusDown peer update.
+//
 // We track dialing and connected states independently. This allows us to accept
 // an inbound connection from a peer while the router is also dialing an
 // outbound connection to that same peer, which will cause the dialer to
 // eventually error (when attempting to mark the peer as connected). This also
 // avoids race conditions where multiple goroutines may end up dialing a peer if
 // an incoming connection was briefly accepted and disconnected while we were
 // also dialing.
-type peerManager struct {
+type PeerManager struct {
+	options PeerManagerOptions
+
 	mtx           sync.Mutex
 	store         *peerStore
 	dialing       map[NodeID]bool
 	connected     map[NodeID]bool
+	evicting      map[NodeID]bool
 	subscriptions map[*PeerUpdatesCh]*PeerUpdatesCh // keyed by struct identity (address)
 }
 
-// newPeerManager creates a new peer manager.
-func newPeerManager(store *peerStore) *peerManager {
-	return &peerManager{
-		store:         store,
+// PeerManagerOptions specifies options for a PeerManager.
+type PeerManagerOptions struct {
+	// PersistentPeers are peers that we want to maintain persistent connections
+	// to. These will be scored higher than other peers, and if
+	// MaxConnectedUpgrade is non-zero any lower-scored peers will be evicted if
+	// necessary to make room for these.
+	PersistentPeers []NodeID
+
+	// MaxConnected is the maximum number of connected peers (inbound and
+	// outbound). 0 means no limit.
+	MaxConnected uint16
+
+	// MaxConnectedUpgrade is the maximum number of additional peer connections to
```

Comment:
Sorry if this is obvious, but is this `MaxConnectedUpgrade` new wrt to the existing p2p options? I'm trying to understand why `MaxConnectedUpgrade` is necessary? If we have exhausted `MaxConnected`, why not just evict a peer with the lowest score from that set? Why have an additional upgrade/buffer?

Reply:
Because if we're full, but we know of better peers, we don't want to evict a lower peer until we've actually tried to dial the new peer, handshaked with it, and made sure it's operational. Only then should we evict the lower peer. Otherwise we'll be evicting peers every time we retry a higher-ranked peer (which may still be offline). While we're doing this, we will for a brief moment be connected to both the old and new peer.

Reply:
> Because if we're full, but we know of better peers, we don't want to evict a lower peer until we've actually tried to dial the better peer, handshaked with it, and made sure it's operational. Only then should we evict the lower peer.

Yup, I understand this. I'm just trying to make the connection with that logic and the use of `MaxConnectedUpgrade`. Can't we still evict and perform that logic and just use `MaxConnected`?

Reply:
`MaxConnectedUpgrade` is basically a cap on the number of probes to better peers. Let's assume we're in a network of 200 nodes (100 high 100 low), with a `MaxConnected` of 50, and we're only connected to low-scored peers. We can't evict before we connect to new nodes. How many better peers should we attempt to connect to simultaneously? All of them?

Reply:
Got it. Thank you.

> MaxConnectedUpgrade is basically a cap on the number of probes to better peers.

This clarified it for me. Maybe add this oneliner to the godoc?

Reply:
Sure!